### ISSUE TITLE
Reading speed and voice control for text-to-speech (TTS) with extended hint (-r)

### ISSUE THREAD

**Issue by gsbabil** [cmcaine/tridactyl#217]
Currently Tridactyl supports some excellent extended hint options
via [these choices][1]. The `-r` or `;r` in particular is very
useful for reading webpages aloud and for accessibility reasons.
External software is often needed otherwise for similar tasks such
as [TextAloud from NextUP.com][2].

Although the TTS hint-mode via `-r` or `;r` is fairly usable at the
moment, it doesn't seem to allow controls of reading speed or
different voices. Similar options are available via Firefox's
[built-in reader-view][3]. It would be very useful if the TTS hint
via `-r` or `;r` either shared the settings from Firefox's built-in
reader-view's TTS settings, or implemented independent reading
speed, voice control etc.

Thanks a lot.

[1]: https://github.com/cmcaine/tridactyl/blob/91d8bd1/src/excmds.ts#L1301
[2]: http://nextup.com/download.html
[3]: https://user-images.githubusercontent.com/1061108/33696599-1cf9b338-db57-11e7-96e8-dc4151ec8e68.png

![reader-view](https://user-images.githubusercontent.com/1061108/33696599-1cf9b338-db57-11e7-96e8-dc4151ec8e68.png)

**Comment by gsbabil** [comment]
At another look it appears that `ttscontrol` provides the necessary
stop/pause functionalities. I was able to get them to work with the
following bindings:

```
:bind ;R ttscontrol stop
:bind ;p ttscontrol pause
```

Thank you!

**Comment by gsbabil** [comment]
Also the TTS settings seem to configurable via [`USERCONFIG`][1]. I
see the following to be the current default:

```
"ttsvoice": "default",  // chosen from the listvoices list, or "default"
"ttsvolume": 1,         // 0 to 1
"ttsrate": 1,           // 0.1 to 10
"ttspitch": 1,          // 0 to 2
```

Could anyone please share the following:

  1. How do I set `USERCONFIG`, or amend `CONFIG` as shown [here][1]?
  2. How do I enlist and modify the my current settings, key
     bindings etc.?

[1]: https://github.com/cmcaine/tridactyl/blob/91d8bd1e39fd5e1e1c04d6724513ad7fc5e3395e/src/config.ts

**Comment by bovine3dom** [comment]
`set [setting] [value]` is the correct syntax (no equals sign).

However, we don't currently support values with spaces in, so there's a bug there (on Linux the voices have no spaces in their names).

I just need to make `set` join the values if the setting is a string.

**Comment by gsbabil** [comment]
Hi @bovine3dom,

Thanks for the clarification on `ttsvoice`.

One more note - I'm currently on `Tridactyl-1.7.0-36-g91d8bd1`. I
just tried setting the `ttsrate` to `10` via `:set ttsrate 10`,
followed by `;r` on a target paragraph. My `ttsvoice` is currently
set to `default` and the `ttsrate` values doesn't seem to have an
impact on the reading speed.

Also, could you please advise a way to enlist my current
self-configured settings in Tridactyl? For example, I'd like to see
all my user-defined `bind` settings at least.

Thanks a lot.

**Comment by bovine3dom** [comment]
I'll look into ttsrate. It worked last time I tried on Linux, so it could be a Windows issue.

The only way to view binds at the moment is a bit hacky: go to `about:debugging` tick "enable addon debugging", scroll down to Tridactyl, click debug, and then type `tri.config.get("nmaps")` into the console that appears.

**Comment by holtzermann17** [comment]
Hi, I was trying to use the following TTS:

```
tri.config.get("ttsvoice")
"\"GoogleStandard British English (Anna)\""
```

But it doesn't work, I get one of the more classic-sounding voices.  However, "Anna" works fine when I select text and press play with this extension: https://addons.mozilla.org/en-GB/firefox/addon/read-aloud/

**Comment by bovine3dom** [comment]
@holtzermann17 the quote marks are a bit suspect, could you try it without them? I.e. `set ttsvoice GoogleStandard British English (Anna)`

**Comment by holtzermann17** [comment]
@bovine3dom yeah, still ends up w/ a robotic sounding voice rather than the dulcet tones of "Anna"

**Comment by bovine3dom** [comment]
What's the output of `:ttsvoices`?

If it's empty, could you open the web console (easiest to right click on a web page, inspect element, then clock the console tab) and run `window.speechSynthesis.getVoices()` and tell us what the result is?

**Comment by holtzermann17** [comment]
The variable doesn't seem to include the GoogleStandard voices on it.

https://gist.github.com/holtzermann17/df0c5daa815d7d4e08c217ed0b3f6c6c

**Comment by bovine3dom** [comment]
Thanks, that's really helpful :)

I did a bit of digging and the extension you linked to is using "the cloud" - https://github.com/ken107/read-aloud/blob/9f36f3c4eeb88f89a85af4b5b37f7a13ecc2e5cd/js/tts-engines.js#L862 - which I do my best to pretend doesn't exist. Tridactyl only supports voices installed locally at the moment.

If you want Tridactyl to support voices like that could you open another issue? We might be able to persuade read-aloud to provide an API for other extensions or refactor their extension into a library that we can import at compile time.
